notebooks/2_uninformed_search/index.md

```diff
@@ -0,0 +1,247 @@
+
+
+## Iterative deepening depth-first search
+
+The iterative deepening algorithm is a combination of DFS and BFS algorithms. This search algorithm finds out the best depth limit and does it by gradually increasing the limit until a goal is found.
+
+This algorithm performs depth-first search up to a certain "depth limit", and it keeps increasing the depth limit after each iteration until the goal node is found.
+
+This Search algorithm combines the benefits of Breadth-first search's fast search and depth-first search's memory efficiency.
+
+The iterative search algorithm is useful uninformed search when search space is large, and depth of goal node is unknown.
+
+**Example**
+
+Look at the example below.
+
+![](https://github.com/mohsenosl99/notes/blob/master/notebooks/2_uninformed_search/images/IDDFS.jpg)
+
+Suppose that node M is goal. So we start from Limit =0 (node A). We check that A is goal or not, if it is not we add 1 to limit and check the nodes of B and C. So continue until reach the goal.
+
+### Performance Measure:
+
+**Completeness**
+
+This algorithm is complete is ifthe branching factor is finite.
+
+**Time Complexity**
+
+> T( b ) = (d+1)b^0 + db^1 + (d−1)b^2 + ... + bd = O( b^d )
+
+or more precisely
+
+> O(b^d(1 – 1/b)^-2)
+
+Note: In this algorithm because of the fact that we want to avoid space problems, we won't store any data therefore we may have to repeat some actions but it won't trouble us because time complexity still remains O( b^d ), similar to BFS.
+
+
+**Space Complexity**
+
+The space complexity of IDDFS will be O(bd).
+
+**Optimality**
+
+IDDFS algorithm is optimal if path cost equals 1.
+
+
+
+
+
+
+## Bidirectional search
+This algorithm is one of the algorithms that is being used for searching a graph or tree. It is originated from BFS algorithm but with the difference that it starts searching from two nodes and precede to search until those two intersect with each other. This algorithm is so much faster and simpler than the BFS algorithm and using this algorithm is appropriate when all the nodes or at least goal states are defined clearly and the branching factor is exactly the same in both directions. For example, in the chart below. If we start from nodes 0 and 14, these two will reach out to each other on node 7 and then the path will be found.
+ 
+![](https://github.com/mohsenosl99/notes/blob/master/notebooks/2_uninformed_search/images/bidir.png)
+
+Now let’s talk about its features and performance measures.
+
+**Completeness**
+
+It is a complete algorithm.
+
+**Optimality**
+
+Again like BFS algorithm, it is not an optimal algorithm in general but if cost of steps be equal to one, this algorithm would be optimal.
+
+**Time and Space Complexity**
+
+It uses memory to save queues and nodes that had been expanded. But the time for reaching an optimal solution is half of the time in BFS algorithm and this is what makes this algorithm so attractive to use.
+
+
+
+## Conclusion
+
+
+
+
+
+
+
+
+
+
```

Comment:
Please Add References, For Example : You Can Add Where You Read Articles About The Topic, or You can mention where students can find more information about the topic